Modules/_ssl.c

```diff
@@ -5514,9 +5534,12 @@ PySSLSession_richcompare(PyObject *left, PyObject *right, int op)
     }
 
     int result;
-    PyTypeObject *sesstype = ((PySSLSession*)left)->ctx->state->PySSLSession_Type;
 
-    if (!Py_IS_TYPE(left, sesstype) || !Py_IS_TYPE(right, sesstype)) {
+    PySSLSession *self = PySSLSession_CAST(left);
+    PyTypeObject *sesstype = Py_TYPE(self);
+    assert(sesstype == self->ctx->state->PySSLSession_Type);
+
+    if (!Py_IS_TYPE(right, sesstype)) {
```

Comment:
I would prefer to keep the left+right checks in this PR. If you want to make the check smarter, please make it in a separated PR.